azkaban-exec-server/src/test/java/azkaban/execapp/FlowRunnerTestBase.java

```diff
@@ -170,4 +171,8 @@ protected void succeedJobs(final String... jobs) {
     }
   }
 
+  protected void assertEvents(final Type... types) {
```

Comment:
Ok, disclaimer first of all: I totally respect the principles that you have. I think it's easier for both of us that I just make the changes that you want, even if my own preferences would be different. Below I'll also answer your questions, because maybe I can learn something here.

> What's the advantage for turning the `eventCollector` from a local variable into a member variable?

That you don't need that extra arg on every single line that needs to assert events.

> prefer composition over inheritance

I recognize the problems with inheritance in general, but sometimes it's just the simplest way to go. In this case I personally don't see that it's causing any real confusion. In case of unit tests I feel more like working on a standalone script than part of bigger application architecture, and there I feel it's better to establish a kind of DSL that makes it easy to create and understand different test cases: easily seeing the essential parts of what's being tested. So I try to remove as much noise as possible.

I checked https://github.com/azkaban/azkaban/pull/1151 and it's fine, but even there every test class has this "much" duplicate code:
```java
this.flowMap = FlowRunnerTestUtil
    .prepareProject(this.project, dir, this.logger, this.workingDir);
```

If you want to add an optional argument or something like that, you need to update every class that calls this method. With inheritance you can just add a new field to the base class and deal with it without having to update the subclasses if they're not interested in that "new feature".

With inheritance all the other args could be protected fields in the base class, and maybe in `setUp` there would be:

```java
super.prepareProject(dir);
```

> avoid redirection

I'm pretty sure `eventCollector` could be created in `FlowRunnerTestBase#setUp` and `FlowRunnerTest#setUp` would call `super.setUp()`. Maybe it could even be `private` then. But maybe that doesn't really improve the structure from your perspective.

> If you can break up this pull request into simpler multiple ones, we can merge the first commit while we discuss the rest. This is another advantage of smaller and independent pull requests i.e. we can treat different parts differently.

That is very true, but will add quite much overhead (on my part) when the changes build upon each other. Any way I will try my best to separate different things into separate PRs.